packages/react-storage/src/components/StorageBrowser/views/LocationActionView/CopyView/useCopyView.ts

```diff
@@ -1,5 +1,95 @@
+import { useState } from 'react';
+
+import { isFunction } from '@aws-amplify/ui';
+
+import { copyHandler } from '../../../actions/handlers';
+import { useProcessTasks } from '../../../tasks';
+import { useGetActionInput } from '../../../providers/configuration';
+import { useStore } from '../../../providers/store';
+import { getActionViewDisabledButtons } from '../utils';
+import { getTaskCounts } from '../../../controls/getTaskCounts';
+import { getDestinationListFullPrefix } from '../utils/getDestinationPickerDataTable';
 import { CopyViewState } from './types';
 
-export function useCopyView(): CopyViewState {
-  throw new Error('useCopyView is not implemented');
-}
+export const useCopyView = ({
+  onExit: _onExit,
+}: {
+  onExit?: () => void;
+}): CopyViewState => {
+  const [
+    {
+      location,
+      locationItems: { fileDataItems },
+    },
+    dispatchStoreAction,
+  ] = useStore();
+  const { key, current } = location;
+
+  const getInput = useGetActionInput();
+
+  const [tasks, handleProcess] = useProcessTasks(
+    copyHandler,
+    // @ts-ignore
+    // temporarily ignore error that items is missing
+    // until PR that refactors useProcessTasks is merged
+    fileDataItems,
+    {
+      concurrency: 1,
+    }
+  );
+
+  const destinationListWithoutSlashes =
+    // handle root bucket access grant
+    key === ''
+      ? ['']
+      : // handle subfolder inside root access grant
+      key && current?.prefix == ''
+      ? ['', ...key.split('/').slice(0, -1)]
+      : // regular access that starts at prefix (not root bucket)
+      key.includes('/')
+      ? key.split('/').slice(0, -1)
+      : [];
+
+  const [destinationList, onSetDestinationList] = useState(
+    destinationListWithoutSlashes
+  );
```

Comment:
The logic in `destinationListWithoutSlashes` should probably be in util with maybe a bit less of a verbose name with it's own unit tests which can then be provided in a callback to the initializer in `useState`, for example:

```
const [destinationList, onSetDestinationList] = useState(
    () => parseDestination(key)
);
```

Reply:
🍕

Reply:
⭐ suck